src/xas_tdp_methods.F

```diff
@@ -0,0 +1,2964 @@
+      TYPE(cp_fm_struct_type), POINTER                   :: ks_struct, lmo_struct
+      TYPE(cp_fm_type), POINTER                          :: evecs, ks_fm, lmo_fm, mo_coeff, work
+      TYPE(cp_para_env_type), POINTER                    :: para_env
+      TYPE(dbcsr_p_type), DIMENSION(:), POINTER          :: matrix_ks
+      TYPE(mo_set_p_type), DIMENSION(:), POINTER         :: mos
+
+      NULLIFY (mos, mo_coeff, matrix_ks, para_env, blacs_env, lmo_struct, lmo_fm, ks_struct, ks_fm)
+      NULLIFY (evecs, work)
+
+      ! Get what we need from qs_env
+      CALL get_qs_env(qs_env, mos=mos, matrix_ks=matrix_ks, para_env=para_env, blacs_env=blacs_env)
+
+      nspins = 1; IF (xas_tdp_control%do_uks) nspins = 2
+
+      ! Loop over the excited atoms and spin
+      DO ispin = 1, nspins
+         DO iat = 1, xas_tdp_env%nex_atoms
+
+            ! get the MOs
+            CALL get_mo_set(mos(ispin)%mo_set, mo_coeff=mo_coeff, nao=nao)
+
+            ! count how many MOs are associated to this atom and create a fm/struct
+            nlmo = COUNT(xas_tdp_env%mos_of_ex_atoms(:, iat, ispin) == 1)
+            CALL cp_fm_struct_create(lmo_struct, nrow_global=nao, ncol_global=nlmo, &
+                                     para_env=para_env, context=blacs_env)
+            CALL cp_fm_create(lmo_fm, lmo_struct)
+            CALL cp_fm_create(work, lmo_struct)
+
+            CALL cp_fm_struct_create(ks_struct, nrow_global=nlmo, ncol_global=nlmo, &
+                                     para_env=para_env, context=blacs_env)
+            CALL cp_fm_create(ks_fm, ks_struct)
+            CALL cp_fm_create(evecs, ks_struct)
+
+            ! Loop over the localized MOs associated to this atom
+            i = 0
+            DO ilmo = 1, xas_tdp_control%n_search
+               IF (xas_tdp_env%mos_of_ex_atoms(ilmo, iat, ispin) == -1) CYCLE
+
+               i = i+1
+               ! put the coeff in our atom-restricted lmo_fm
+               CALL cp_fm_to_fm_submat(mo_coeff, lmo_fm, nrow=nao, ncol=1, s_firstrow=1, &
+                                       s_firstcol=ilmo, t_firstrow=1, t_firstcol=i)
+
+            END DO !ilmo
+
+            ! Computing the KS matrix in the subset of MOs
+            CALL cp_dbcsr_sm_fm_multiply(matrix_ks(ispin)%matrix, lmo_fm, work, ncol=nlmo)
+            CALL cp_gemm('T', 'N', nlmo, nlmo, nao, 1.0_dp, lmo_fm, work, 0.0_dp, ks_fm)
+
+            ! Diagonalizing the KS matrix in the subset of MOs
+            ALLOCATE (evals(nlmo))
+            CALL cp_fm_syevd(ks_fm, evecs, evals)
+            DEALLOCATE (evals)
+
+            ! Express the MOs in the basis that diagonalizes KS
+            CALL cp_gemm('N', 'N', nao, nlmo, nlmo, 1.0_dp, lmo_fm, evecs, 0.0_dp, work)
+
+            ! Replacing the new MOs back in the MO coeffs
+            i = 0
+            DO ilmo = 1, xas_tdp_control%n_search
+               IF (xas_tdp_env%mos_of_ex_atoms(ilmo, iat, ispin) == -1) CYCLE
+
+               i = i+1
```

Comment:
so, `i` is actually `ilmo-1`?

Reply:
Not quite. Because of the `IF` condition above, `ilmo` and `i` might evolve at different rates.

Reply:
ah, that one I overlooked, sorry